device/core/src/client_diagnostic.ts

```diff
@@ -0,0 +1,205 @@
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_002: [This constructor shall set message number to 0.]
+    this.currentMessageNumber = 0;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.getDiagSamplingPercentage
+   * @description       get the diagnostic sampling percentage.
+   *
+   * @returns {number}
+   */
+  getDiagSamplingPercentage(): number {
+    return this.diagSamplingPercentage;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.setDiagSamplingPercentage
+   * @description       set the diagnostic sampling percentage.
+   *
+   * @param {number}   diagSamplingPercentage   The value of sampling percentage.
+   * @throws {ArgumentError}     If value < 0 or value > 100.
+   */
+  setDiagSamplingPercentage(diagSamplingPercentage: number): void {
+    if (typeof diagSamplingPercentage !== 'number') {
+      throw new errors.ArgumentError('Sampling percentage only accept number');
+    }
+    if (diagSamplingPercentage % 1 !== 0) {
+      throw new errors.ArgumentError('Sampling percentage should be integer');
+    }
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_003: [When percentage is less than 0 or larger than 100, throw IllegalArgumentException.]
+    if (diagSamplingPercentage < 0 || diagSamplingPercentage > 100) {
+      throw new errors.ArgumentError('Sampling percentage should be [0,100]');
+    }
+    this.diagSamplingPercentage = diagSamplingPercentage;
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_004: [This function shall reset message number to 0.]
+    this.currentMessageNumber = 0;
+  }
+
+  /**
+   * @method            module:azure-iot-device.DiagnosticClient.addDiagnosticInfoIfNecessary
+   * @description       add the diagnostic info
+   *
+   * @param {Message}   message   The message
+   */
+  addDiagnosticInfoIfNecessary(message: Message): void {
+    if (this.shouldAddDiagnosticInfo()) {
+      let diagnosticPropertyData: DiagnosticPropertyData = new DiagnosticPropertyData(this.generateEightRandomCharacters(), this.getCurrentTimeUtc());
+      message.diagnosticPropertyData = diagnosticPropertyData;
+    }
+  }
+
+  // Get a character from 0-9a-zA-Z
+  private getDiagnosticIdChar(value: number): string {
+    if (value <= 9) {
+      // 48 is char code of '0'
+      return String.fromCharCode(48 + value);
+    } else if (value <= 9 + 26) {
+      // 65 is char code of 'A'
+      return String.fromCharCode(65 + value - 10);
+    } else {
+      // 97 is char code of 'a'
+      return String.fromCharCode(97 + value - 36);
+    }
+  }
+
+  private generateEightRandomCharacters(): string {
+    // Codes_SRS_DEVICECLIENTDIAGNOSTIC_01_005: [This function shall generate 8 random chars, each is from 0-9a-z.]
+    let result: string = '';
+    for (let i = 0; i < 8; i++) {
+      let randomNum: number = Math.floor(Math.random() * DiagnosticClient.DIAGNOSTIC_ID_CHARACTER_BASE);
+      result += this.getDiagnosticIdChar(randomNum);
+    }
+    return result.toString();
+  }
```

Comment:
or we could just use guids and save us all that trouble.

Reply:
We discussed this with IoT Hub team and made the decision.
You can refer to our design doc here: https://teams.microsoft.com/_#/docx/viewer/teams/https%3A~2F~2Fmicrosoft.sharepoint.com~2Fteams~2FIoTToolingTeam~2FShared%20Documents~2FEnd-to-End%20Diagnostics~2FAzure%20IoT%20Hub%20Device%20Client%20SDKs%20-%20E2E%20Diagnostic%20P0%20scenario.docx?threadId=19%3Acecb83aeeeea41bb818083fc32b545bd%40thread.skype&baseUrl=https%3A~2F~2Fmicrosoft.sharepoint.com~2Fteams~2FIoTToolingTeam&fileId=03F735E4-E751-4917-A3E4-EF04CFB2AAD6&ctx=files&viewerAction=view

Reply:
I do not have access to this

---
In reply to: [155441477](https://github.com/Azure/azure-iot-sdk-node/pull/158#discussion_r155441477) [](ancestors = 155441477)